### ISSUE TITLE
Increase Aliases/Grouping to handle more complex networks.

### ISSUE THREAD

**Issue by stevecs** [opnsense/core#6788]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [ X ] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [ X ] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Is your feature request related to a problem? Please describe.**

When trying to organize and build out complex policies the Alias name function should be expanded to 

- allow longer names (>32 characters)
- and other characters than [0-9,a-z,A-Z,_] such as '.' ':' '/' so as to encompass IPv4/v6 objects
- Allow exclusions

**Describe the solution you like**

When creating policies with hundreds/thousands of objects (i.e. complex sites with multiple DMZ's, or extended sites where there may be feeder sites, central sites, multiple egress locations (different offices with internal MPLS but each having their own Inet egress and each having local DMZ's) and trying to maintain a consistent policy structure across them (so as to not have 10's or hundreds of one-off policies but have a consistent object/alias database which can be applied to *all* firewalls and then be used to construct a generic policy to be applied to similar enforcement points (i.e. all associate firewalls have X policy, all HR firewalls have Y policy, all application FOO have Z policies).  

In 'enterprise' firewalls (Checkpoint, Palo, Cisco, etc).  You can do this by creating objects such as:
Hst_a.b.c.d
Net_a.b.c.d_XX  (Net_a.b.c.d/XX) where XX is the bitmask of the network (assuming IPv4 here but similar concept for IPv6)

Which then can be rolled into aliases/groups which can already be done in opnsense which then can be used in policies/rules.  So that the host/network is *never* used directly in a rulebase to provide abstraction.  I.e.  If you have say 3 sites a DMZ each with say and all DMZ's are extended via mpls to each other

Net_10.1.1.0_24
Net_10.1.254.240_28 (mpls/wan router site 1)
Net_10.2.1.0_24
Net_10.2.254.240_28 (mpls/wan router site 2)
Net_10.3.1.0_24
Net_10.3.254.240_28 (mpls/wan router site 3)

This can be rolled into group aliases such as:
Company_App_Site1_DMZ  (contains above site1 _24's) 
Company_App_Site1_Rtr (contains above site1 _28's)
Company_App_Site2_DMZ (contains above site2 _24's)
Company_App_Site2_Rtr (contains above site2 _28's) 

Which you can then roll up into say:
Company_App  (i.e. contains all _SiteX_) items which can be used in a policy.

The exclusions come in when you are looking to create anti-spoofing and similar rules.  I.e.  When applying rules to an interface (not floating) where if you have each individual firewall needing to know what is 'local' and what is 'remote' and distinguish that from 'internet'.     I.e.  

DMZ1 - (extended DMZ across multiple sites as above with collection of networks)
CorpWAN - (defined by all other networks that traffic needs to get to that is NOT locally tied and NOT internet)
WAN/Inet - (defined by what is NOT defined in the alias database.  i.e.  it cannot be 0.0.0.0/0 as that *includes* internal addresses which is not correct,  It needs to be defined by 0.0.0.0/0 MINUS what you have internally) 

So an exclusion group is needed for anti-spoofing and policy/rule control.   I.e.  
Alias: SG_SITE1_CorpWAN 
    would contain a roll-up or all networks (Company_*) with EXCLUSION of whatever Company_*_Site1_* networks would be.  So if there happens to be a route leak from mpls or other screw-up, if a frame comes in from the CorpWAN interface but it is from a network that should be on the SiteX_DMZ interface it would be blocked/detected.  (i.e. lets you find/control routing loop or asymmetric  routing problems).

Alias SG_Inet
    would contain a simple version  of 0.0.0.0/0 with EXCLUSION of Company_* aliases (i.e. all networks inside company from any internally connected site).

EXCLUSIONS I can see being implemented on the alias screen similar to the 'content' add function but to have an 'exclusion' version of that and the resultant group/alias would be the result of those two arrays.    This will require some logic/expansion as for example if you have objects that are say 172.18.0.0/24 and 172.16.0.0/12 or overlapping bitmasks you would need to create a dynamic map/array of the resultant networks.

Lastly, this is also important when creating VPN tunnels to other sites as you need to have your encryption domains to be unique to each site (or you would never enter the tunnels).   Within your own company you can do this by back-end routers and just create a GRE tunnel (so encryption domain is just the /32's at each site) but that doesn't work well when dealing with vendors/partners or when merging/separating companies. 

Basically the above *CAN* be done w/ OPNsense with basically two changes

- Expand the size & allowed characters in the aliases array/database
- Allow for the expansion of those objects and create an exclusion/negation of a subset of objects.  

This may be a little out of scope of the opnsense target market, I come from enterprise worlds and see the benefit here to use community projects both for home, soho, and corporate but all such projects seem to be lacking in similar fashions on dealing with complex networking and would like to see that chasm shrink a bit.

**Comment by fichtner** [comment]
Groups are pinned between interface < 16 chars and pf alias variable name (underscore, alphanum) constraint. Doing something else would require to be interface names and alias variables to be unreadable for humans, and humans do complain when they are not as the past has told us. 😉

**Comment by stevecs** [comment]
The above is human readable, or I'm not following what you're trying to say.   I've deployed hundreds of firewalls with similar structures to handle complex environments.  

Are not aliases/groups already an abstraction (i.e. name field in a sql database)?

**Comment by fichtner** [comment]
I’ve mentioned both OS restrictions. I do t doubt the utility of the ideas voiced here but as I said we are restricted by the OS.

**Comment by AdSchellevis** [comment]
Exclusions https://docs.opnsense.org/manual/aliases.html#exclusions , https://docs.freebsd.org/en/books/handbook/firewalls/#pftut-tables ?

**Comment by stevecs** [comment]
I guess I am still not following when you mention 'both OS limitations'.    When creating an alias I am talking about the NAME field here.    I do not understand where you say that is an OS limit, can you point to where that is used by the OS directly opposed to being stored in a database?   If that is used as a FILENAME that would be very limited and would seem that community based firewalls have a lot further to go than I thought.

**Comment by fichtner** [comment]
Why don’t you trust me? Do you know anything that I don’t?

**Comment by stevecs** [comment]
Well first of all, I don't 'trust' anyone.  I've been doing security for nearly 40 years.   :)  but that's immaterial I am looking to understand where the issue is.    I am not following what you're saying there the 'name' is a limitation of the OS.

From the other link that AdSchellevis provided it appears that the 'content' is being used directly (i.e. not parsed) which I can see the exclusion part of what I was suggesting originally being much more involved (i.e. can't be used directly and would have to re-write the array of networks whenever the object was updated which can take time to 'compile' such a list..

**Comment by AdSchellevis** [comment]
`man pf(4)` for the name part, but you can use any description you like (semantics). No clue what you mean with not being parsed, but I'm pretty sure you can find what you need in our documentation (the nesting part for example how to stack concepts).
